
<div style="text-align:center">

<div style="background-image: url('background.jpg'); background-size: cover; padding: 50px; color: white;">

# Acoustic Loggers 

© Sand Technologies

<img src="sand.JPEG" alt="Example Image" width="100" height="100">

### Team 1

</div>
</div>


### **Acoustic Loggers for Leak Detection**


Water distribution networks play a crucial role in ensuring clean and safe drinking water is delivered to consumers. However, leaks in the pipes of these networks lead to significant loss of water, posing challenges to water utilities. Not only are these leaks wasteful, but they also result in large fines for water utilities. 

To mitigate wastage, acoustic loggers have been attached to water pipes to record the sound profile in each pipe at night. These recordings can then be used to determine whether there is a leak present. 

The goal of this project is to produce a model that can classify each of these recordings as either 'leak' or 'no leak', aiding in the early detection and prevention of water loss in distribution networks. 

<img src="water.JPG" alt="Example Image" width="1700" height="400">

### **Problem Statement:**




<a id="cont"></a>

## Table of Contents

<a href=#one>1. Importing Libraries</a>

<a href=#two>2. Loading the data</a>

<a href=#four>4. Exploratory Data Analysis</a>

<a href=#five>5. Data Cleaning & Preprocessing

<a href=#six>6. Modelling</a>

<a href=#seven>7. Model Evaluation</a>

<a href=#ten>10. Conclusion</a>

## <div style="text-align: center;"><u/> **Let's Get Started!!!**.</u></div>

 <a id="two"></a>
## **1. Importing Libraries**
<a href=#cont>Back to Table of Contents</a>

---


In [4]:
!pip install requests
!pip install librosa
import requests
import zipfile
import os
import pandas as pd
import os
import io
import matplotlib.pyplot as plt
import numpy as np
import librosa
import librosa.display
import IPython.display as ipd
import boto3
from io import BytesIO


[notice] A new release of pip is available: 23.3.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


   ---------------------------------------- 0.0/253.7 kB ? eta -:--:--
   ---------------------- ----------------- 143.4/253.7 kB 4.3 MB/s eta 0:00:01
   ---------------------------------------- 253.7/253.7 kB 3.9 MB/s eta 0:00:00
   ---------------------------------------- 0.0/75.2 kB ? eta -:--:--
   ---------------------------------------- 75.2/75.2 kB 4.3 MB/s eta 0:00:00
   ---------------------------------------- 0.0/2.6 MB ? eta -:--:--
   ------- -------------------------------- 0.5/2.6 MB 15.5 MB/s eta 0:00:01
   ------------------------- -------------- 1.7/2.6 MB 21.4 MB/s eta 0:00:01
   ---------------------------------------  2.6/2.6 MB 20.9 MB/s eta 0:00:01
   ---------------------------------------- 2.6/2.6 MB 18.7 MB/s eta 0:00:00
   ---------------------------------------- 0.0/63.0 kB ? eta -:--:--
   ---------------------------------------- 63.0/63.0 kB ? eta 0:00:00
   ---------------------------------------- 0.0/1.0 MB ? eta -:--:--
   -------------------------------


[notice] A new release of pip is available: 23.3.2 -> 24.0
[notice] To update, run: python.exe -m pip install --upgrade pip


 <a id="three"></a>
## **2. Downloading Data**
<a href=#cont>Back to Table of Contents</a>

---


#### **3.1 About the dataset**

**<u/>Dataset Description</u>**

**Data Overview**.

The dataset comprises unstructured audio files spanning from 2017 to 2022, accompanied by corresponding metadata in an Excel spreadsheet. These audio files capture the sound of water flow within pipes, distinguishing between those with leaks and those without. Themas Waters has graciously provided us access to this data via an API..

**Source**.

The Excel spreadsheet contains a comprehensive set of columns or features, serving as metadata for each audio file as follows:


- **datetime:** This column represents the date and time when the audio recording was captured.

- **siteid:** The site ID is a unique identifier assigned to each location where the recording took place. It helps in tracking the geographical location associated with the audio data.

- **recording_id:** The recording ID is a unique identifier assigned to each audio recording. It distinguishes one recording from another and aids in organizing and referencing the audio files.

- **file_name:** The file name column contains the name of the audio file. It helps in identifying and accessing the corresponding audio recording file.

- **postcodedistrict:** This column contains the postal code district associated with the location where the recording was made. It provides additional geographical context to the data.

- **dmacode:** The DMA (Distribution Management Area) code is a unique identifier used in water management systems. It helps in categorizing and managing water distribution networks.

- **leak found:** This column indicates whether a leak was detected in the corresponding audio recording. It serves as a binary flag, where "leak found" signifies the presence of a leak, and "no leak found" indicates the absence of a leak.

- **noise:** The noise column represents the characteristics of the recorded sound, providing insights into the acoustic properties of the audio data.

- **spread:** The spread column refer to the spread or distribution of sound frequencies within the audio recording. It could provide information about the variability or range of sound frequencies captured in the recording.

#### **3.2 Downloading files from the API**

- API Credentials

In [ ]:
clientID = 'c70b57fc939d4c4eb3b32bc256fe451f'
clientSecret = '515600b3BB9547A580760B29007c6E73'

- modify this url as desired to access the different end points. e.g. 

In [ ]:
# Replace DischargeCurrentStatus at the end of the resource URL
api_root = 'https://prod-tw-opendata-app.uk-e1.cloudhub.io'
api_resource = '/data/AcousticLogger/v1/SoundFiles'
url = api_root + api_resource
params = 'data filters' # Parameter

- Reqesting Data from the URL

In [ ]:
r = requests.get(url, headers={'client_id':clientID, 'client_secret': clientSecret}, params=params)
print("Requesting from " + r.url)

- Checking request status to validate the request.

In [ ]:
if r.status_code == 200:
    response = r.json()
    df = pd.json_normalize(response, 'items')
else:
    raise Exception("Request failed with status code {0}, and error message: {1}".format(r.status_code, r.json()))

- We have retrieved the data Lets take what we want from this data.

In [4]:
print(df.tail())
a = df.loc[0, 'FileURL']
a
response = requests.get(a)

save_path = r'C:\Users\Percy\OneDrive\Desktop\Acoustic\0404'

if response.status_code == 200:
    with io.BytesIO(response.content) as zip_data:
        
        with zipfile.ZipFile(zip_data, 'r') as zip_ref:
            zip_ref.extractall(os.path.abspath(r'C:\Users\Percy\OneDrive\Desktop\Acoustic\0404'))  # Specify the destination folder
                
    print("Zip file downloaded successfully.")
else:
    print(f"Failed to download zip file. Status code: {response.status_code}")

Requesting from https://prod-tw-opendata-app.uk-e1.cloudhub.io/data/AcousticLogger/v1/SoundFiles?data%20filters
            FileName                                            FileURL  \
0  acoustic_data.zip  https://saseuwdevdsplat.blob.core.windows.net/...   

                        Date  
0  2024-02-06T11:34:43+00:00  
Zip file downloaded successfully.


<a id="four"></a>
## **4. Connecting Audio with Metadata**
<a class="anchor" id="1.1"></a>
<a href=#cont>Back to Table of Contents</a>

---
    

#### **4.1 Retrieving metadata.**

4.1.1 <u/>Lets take the metadeta from the s3 bucket into the notebook</u>

In [25]:
s3 = boto3.client('s3')
bucket_name = '2307-01-acoustic-loggers-for-leak-detection-a'
object_key = 'Metadata/labelled_acoustic_logger_leaks.xlsx'

response = s3.get_object(Bucket=bucket_name, Key=object_key)
excel_data = response['Body'].read()

metadata_df = pd.read_excel(BytesIO(excel_data))
print('Metadata Loaded Successfully into the notebook')

Metadata Loaded Successfully into the notebook


4.1.2 <u/>Lets Observe the first 5 raws of the metadata</u>

In [24]:
metadata_df.head()

datetime   siteid  recording_id  \
0 2018-12-12 04:00:00  1668120      26577010   
1 2018-12-12 04:00:00  1742872      26592074   
2 2018-12-12 04:00:00  1616760      26593071   
3 2018-12-12 04:00:00  1630929      26593758   
4 2018-12-12 04:00:00  6896951      26596303   

                                                file postcodedistrict  \
0  ../recordings/2018/12/12/recordings_1668120_26...             NW10   
1  ../recordings/2018/12/12/recordings_1742872_26...              E15   
2  ../recordings/2018/12/12/recordings_1616760_26...             HP12   
3  ../recordings/2018/12/12/recordings_1630929_26...              SL1   
4  ../recordings/2018/12/12/recordings_6896951_26...              SL7   

     dmacode leak_found  noise  spread repaired_as  
0   ZSUHIL25         No   15.0     4.0         NaN  
1  ZWOODF113         No   22.0     9.0         NaN  
2   ZWIDDN02         No   21.0     7.0         NaN  
3   ZSTKWD30         No   14.0     5.0         NaN  
4   ZMARLC01         No   10.0     3.0         NaN

#### **4.2 Retrieving Audio.**

4.2.1 <u/>Adding Feature in metadata table that will help us retrieve the audios</u>

In [26]:
metadata_df['year'] = metadata_df['datetime'].dt.year
metadata_df['month'] = metadata_df['datetime'].dt.month
metadata_df['day'] = metadata_df['datetime'].dt.day

4.2.2 <u/>Retrieving list of Audio URL and audio Keys</u>

In [27]:
url = []
key_ = []
for i in range(len(metadata_df)):
    year = metadata_df.loc[i, 'year']
    month = metadata_df.loc[i, 'month']
    day = metadata_df.loc[i, 'day']
    sideid = metadata_df.loc[i, 'siteid']
    rec_id =metadata_df.loc[i, 'recording_id']
    file = metadata_df.loc[1,'file']
    value = file[62:-4]
    
    
    aws_access_key_id = 'AKIATNJHRXAPQBHVQARV'
    aws_secret_access_key = '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'
    region_name = 'eu-west-1'

    # Initialize S3 client
    s3 = boto3.client('s3', 
                     aws_access_key_id=aws_access_key_id, 
                     aws_secret_access_key=aws_secret_access_key, 
                     region_name=region_name)

    # S3 bucket and object details
    bucket_name = '2307-01-acoustic-loggers-for-leak-detection-a'
    key = f'Unstructured audio files/{year}/{month}/{day}/recordings_{sideid}_{rec_id}_{year}{month}{day}_{value}.wav'  # Path to the audio file in the S3 bucket

    key_url = f'Unstructured+audio+files/{year}/{month}/{day}/recordings_{sideid}_{rec_id}_{year}{month}{day}_{value}.wav'
    # Generate a URL to the audio file
    link = f"https://{bucket_name}.s3.{region_name}.amazonaws.com/{key_url}"
    link = [link]
    url = url + link
    key = [key]
    key_ = key_ + key

4.2.3 <u/>Connecting Metadata with corresponding Audio links</u>

In [28]:
metadata_df['Audio_Links'] = url
metadata_df['Audio_key'] = key_

4.2.4 <u/>Observing the first 5 raws of the connected metadata with its audio data</u>

In [29]:
metadata_df.head()

datetime   siteid  recording_id  \
0 2018-12-12 04:00:00  1668120      26577010   
1 2018-12-12 04:00:00  1742872      26592074   
2 2018-12-12 04:00:00  1616760      26593071   
3 2018-12-12 04:00:00  1630929      26593758   
4 2018-12-12 04:00:00  6896951      26596303   

                                                file postcodedistrict  \
0  ../recordings/2018/12/12/recordings_1668120_26...             NW10   
1  ../recordings/2018/12/12/recordings_1742872_26...              E15   
2  ../recordings/2018/12/12/recordings_1616760_26...             HP12   
3  ../recordings/2018/12/12/recordings_1630929_26...              SL1   
4  ../recordings/2018/12/12/recordings_6896951_26...              SL7   

     dmacode leak_found  noise  spread repaired_as  year  month  day  \
0   ZSUHIL25         No   15.0     4.0         NaN  2018     12   12   
1  ZWOODF113         No   22.0     9.0         NaN  2018     12   12   
2   ZWIDDN02         No   21.0     7.0         NaN  2018     12   12   
3   ZSTKWD30         No   14.0     5.0         NaN  2018     12   12   
4   ZMARLC01         No   10.0     3.0         NaN  2018     12   12   

                                         Audio_Links  \
0  https://2307-01-acoustic-loggers-for-leak-dete...   
1  https://2307-01-acoustic-loggers-for-leak-dete...   
2  https://2307-01-acoustic-loggers-for-leak-dete...   
3  https://2307-01-acoustic-loggers-for-leak-dete...   
4  https://2307-01-acoustic-loggers-for-leak-dete...   

                                           Audio_key  
0  Unstructured audio files/2018/12/12/recordings...  
1  Unstructured audio files/2018/12/12/recordings...  
2  Unstructured audio files/2018/12/12/recordings...  
3  Unstructured audio files/2018/12/12/recordings...  
4  Unstructured audio files/2018/12/12/recordings...

**Comment:** The train dataset has 4 features, let's see if they are no missing values withing each feature.

In [30]:
eda_df = metadata_df.copy

<a id="four"></a>
## **6. Exploratory Data Analysis**
<a class="anchor" id="1.1"></a>
<a href=#cont>Back to Table of Contents</a>

---

#### **4.1 Audio Sound.**

6.1.1 <u/>Sound With leak</u>

In [61]:
Leaking = metadata_df[metadata_df['leak_found'] == 'Yes']
Leaking.tail(10)


datetime   siteid  recording_id  \
38320 2020-05-15 03:45:00  6911838      74461571   
38321 2020-08-25 03:45:00  1631655      88529579   
38322 2020-09-03 03:45:00  1659017      89701705   
38323 2021-03-23 03:45:00  1742660     115371709   
38324 2019-07-12 03:45:00  1686827      41327108   
38325 2021-06-23 02:30:00  1769019     129718949   
38326 2020-05-14 03:45:00  7017095      74345861   
38327 2020-05-29 03:45:00  1613699      76414165   
38328 2021-09-16 02:30:00  1722290     143471783   
38329 2021-01-02 03:45:00  1771081     104929501   

                                                    file postcodedistrict  \
38320  ../recordings/2020/05/15/recordings_6911838_74...              SL1   
38321  ../recordings/2020/08/25/recordings_1631655_88...              SL1   
38322  ../recordings/2020/09/03/recordings_1659017_89...             SW15   
38323  ../recordings/2021/03/23/recordings_1742660_11...              E17   
38324  ../recordings/2019/07/12/recordings_1686827_41...              CR2   
38325  ../recordings/2021/06/23/recordings_1769019_12...               E7   
38326  ../recordings/2020/05/14/recordings_7017095_74...              CR0   
38327  ../recordings/2020/05/29/recordings_1613699_76...              SL7   
38328  ../recordings/2021/09/16/recordings_1722290_14...              E14   
38329  ../recordings/2021/01/02/recordings_1771081_10...               E7   

        dmacode leak_found  noise  spread   repaired_as  year  month  day  \
38320  ZSTKWD36        Yes   25.0     9.0           NaN  2020      5   15   
38321  ZSTKWD36        Yes   21.0     6.0           NaN  2020      8   25   
38322  ZHAM3639        Yes   27.0     4.0           NaN  2020      9    3   
38323  ZCHIGW56        Yes   73.0     3.0           NaN  2021      3   23   
38324  ZRUSSH04        Yes   31.0     4.0       Ferrule  2019      7   12   
38325  ZWOODF39        Yes   30.0    10.0  Boundary Box  2021      6   23   
38326  ZRUSSH10        Yes   22.0     7.0  Service Pipe  2020      5   14   
38327  ZMARLT03        Yes   20.0     5.0  Boundary Box  2020      5   29   
38328  ZFINSB56        Yes   25.0     5.0          Main  2021      9   16   
38329  ZWOODF39        Yes   24.0     7.0  Service Pipe  2021      1    2   

                                             Audio_Links  \
38320  https://2307-01-acoustic-loggers-for-leak-dete...   
38321  https://2307-01-acoustic-loggers-for-leak-dete...   
38322  https://2307-01-acoustic-loggers-for-leak-dete...   
38323  https://2307-01-acoustic-loggers-for-leak-dete...   
38324  https://2307-01-acoustic-loggers-for-leak-dete...   
38325  https://2307-01-acoustic-loggers-for-leak-dete...   
38326  https://2307-01-acoustic-loggers-for-leak-dete...   
38327  https://2307-01-acoustic-loggers-for-leak-dete...   
38328  https://2307-01-acoustic-loggers-for-leak-dete...   
38329  https://2307-01-acoustic-loggers-for-leak-dete...   

                                               Audio_key  
38320  Unstructured audio files/2020/5/15/recordings_...  
38321  Unstructured audio files/2020/8/25/recordings_...  
38322  Unstructured audio files/2020/9/3/recordings_1...  
38323  Unstructured audio files/2021/3/23/recordings_...  
38324  Unstructured audio files/2019/7/12/recordings_...  
38325  Unstructured audio files/2021/6/23/recordings_...  
38326  Unstructured audio files/2020/5/14/recordings_...  
38327  Unstructured audio files/2020/5/29/recordings_...  
38328  Unstructured audio files/2021/9/16/recordings_...  
38329  Unstructured audio files/2021/1/2/recordings_1...

In [53]:
audio_file_key = Leaking.loc[22105, 'Audio_key']

In [55]:
# S3 bucket name and audio file path
bucket_name = '2307-01-acoustic-loggers-for-leak-detection-a'
audio_file_key = Leaking.loc[22105, 'Audio_key']  # Path to the audio file in the S3 bucket

# Download the audio file from S3
response = s3.get_object(Bucket=bucket_name, Key=audio_file_key)
audio_data = response['Body'].read()

# Display the audio file using the IPython Audio widget
Audio(audio_data)

NoSuchKey: An error occurred (NoSuchKey) when calling the GetObject operation: The specified key does not exist.

6.1.2 <u/>Sound With No leak</u>

In [58]:
Not_Leaking = metadata_df[metadata_df['leak_found'] == 'No']
Not_Leaking.head()

datetime   siteid  recording_id  \
0 2018-12-12 04:00:00  1668120      26577010   
1 2018-12-12 04:00:00  1742872      26592074   
2 2018-12-12 04:00:00  1616760      26593071   
3 2018-12-12 04:00:00  1630929      26593758   
4 2018-12-12 04:00:00  6896951      26596303   

                                                file postcodedistrict  \
0  ../recordings/2018/12/12/recordings_1668120_26...             NW10   
1  ../recordings/2018/12/12/recordings_1742872_26...              E15   
2  ../recordings/2018/12/12/recordings_1616760_26...             HP12   
3  ../recordings/2018/12/12/recordings_1630929_26...              SL1   
4  ../recordings/2018/12/12/recordings_6896951_26...              SL7   

     dmacode leak_found  noise  spread repaired_as  year  month  day  \
0   ZSUHIL25         No   15.0     4.0         NaN  2018     12   12   
1  ZWOODF113         No   22.0     9.0         NaN  2018     12   12   
2   ZWIDDN02         No   21.0     7.0         NaN  2018     12   12   
3   ZSTKWD30         No   14.0     5.0         NaN  2018     12   12   
4   ZMARLC01         No   10.0     3.0         NaN  2018     12   12   

                                         Audio_Links  \
0  https://2307-01-acoustic-loggers-for-leak-dete...   
1  https://2307-01-acoustic-loggers-for-leak-dete...   
2  https://2307-01-acoustic-loggers-for-leak-dete...   
3  https://2307-01-acoustic-loggers-for-leak-dete...   
4  https://2307-01-acoustic-loggers-for-leak-dete...   

                                           Audio_key  
0  Unstructured audio files/2018/12/12/recordings...  
1  Unstructured audio files/2018/12/12/recordings...  
2  Unstructured audio files/2018/12/12/recordings...  
3  Unstructured audio files/2018/12/12/recordings...  
4  Unstructured audio files/2018/12/12/recordings...

In [59]:
# S3 bucket name and audio file path
bucket_name = '2307-01-acoustic-loggers-for-leak-detection-a'
audio_file_key = Not_Leaking.loc[0, 'Audio_key']  # Path to the audio file in the S3 bucket

# Download the audio file from S3
response = s3.get_object(Bucket=bucket_name, Key=audio_file_key)
audio_data = response['Body'].read()

# Display the audio file using the IPython Audio widget
Audio(audio_data)

4.1.2 <u/>Missing values in Train dataset </u>

**Comment:** Now we can see that they are no missing values within each column, from this dataframe lets see the length of this dataframe, and observe the timber of distinct userid and movieid.

4.1.3 <u/>length of the dataframe and Number of distinct element on movieid and userid </u>

**Comment:** Now we can see that they are more users that movies, let's check the visual to see the frequency of user ratings for the movies. 

4.1.4 <u/>Frequeacy of rating categories </u>

**Comment:** From the above distribution we can see that the rating of 4 makes 26% of the total dataset, this tell us that people rate most of the movies with a value of 4. Now lest observe the statistical information of the given ratings.

4.1.5 <u/>Rating Statistical summary</u>

**Comment:** From the above we see that avarage values is 3.533, and the maximum rating is 5. Let's visualize this knowledge so that we can see the insights of the ratings feature.

4.1.6 <u/>Ratings Visuals</u>
- Box and Whiskers plot.

**Comment:** The skewness of the ratings is -0.7 which tells us that the distribution is left-skewed (tail to the left).

#### **4.2 Movie Dataset Exploratory Data Analysis.**

4.2.1 <u/>The first 3 rows of the Movies dataset </u>

**Comment:** From the movie dataset we can see that the un explored features are text data, lets see if they are missing values.

4.2.2 <u/>Nan values in movies dataset</u>

** Comment:** from the above results it seems like they are no missing values for the movies dataset. now lets see the frequency of movie genres.

4.2.3 <u/>Frequency of genres in movies dataset.</u>

**Comment:** we can see that drama, comedy, documentary, are the top 3 genres. Now lets visualize the frequent genre in a word cloud.

4.2.4 <u/>Frequent Genre.</u>
- Word Cloud visualization.

**Comment:** From the word cloud we can see that drama, Scifi, Romance and comedy are the most frequent genres of our dataset.Lets look at the distribution.

4.2.5 <u/>Genres Distributions</u>

**Comment:** The above distribution confirms that most of the movies in the dataset are drama and romance movies. 

#### **4.3 Tag Dataset Exploratory Data Analysis.**

4.3.1 <u/>The first 3 rows of the Tag dataset </u>

**Comment:** We can see that the dataframe has two new features which is the tag and the timestamp, but for this project we will be interested in the tag which is what was provided by the user.

4.3.2 <u/> Lets check for missing values</u>

**Comment:** We can see that they are no missing values in this dataset. Lets see the cloud of how people classify the genres on the movies.

4.3.3 <u/>Tag genres of people</u>

**Comment:** From what viewers have tagged the movies they have watched, the wordcloud shows that they classify most movies as Comedy and Drama which is what the movies dataset shows. 

#### **4.4 Genome_scores Dataset Exploratory Data Analysis.**

4.4.1 <u/>The first 3 rows of the Genome_scores dataset </u>

**Comment:** From the genome_scores dataset we cans see that the new feature is the relevance of the column. lets check missing values and perform somestatistical analysis on the relevence column.

4.4.2 <u/>Nan values in the geneme_scores dataset<u/>

**Comment:** It can be seen that they are no missing values within the genome score dataset.

4.4.3 <u/>Statistical Sescription<u/>

**Comment:** From the above we can see the mean, maximum and the minimum relevance, lets put this values on a plot to see them in a visual form.

4.4.4 <u/>Statistical summary Plot<u/> 
- Box and whisker plot.

**Comment:** The skewness of the relevence is 2.6 which tells us that the distribution is right-skewed (tail to the right).

#### **4.5 Rating Dataset Exploratory Data Analysis.**

4.5.1 <u/>The first 3 rows of the rating dataset </u>

**Comment:** Since we have already done some analysis on the userid, movieid and ratings features lets only check for missing values.

5.5.2 <u/>Nan valeus<u/>

**Comment:** As we can see they are no missing values that are present in the ratings data frame and all features has been explored, now we jump to the next dataframe.

#### **4.6 imdb Dataset Exploratory Data Analysis.**

4.6.1 <u/>The first 3 rows of the imbd dataset </u>

**Comment:** In this dataset they are 5 new features which we havent explored. lets explore them and see their insights. 

4.6.2 <u/> Statistical Summary<u/>

<a id="five"></a>
## **5. Data Cleaning & Preprocessing**
<a class="anchor" id="1.1"></a>
<a href=#cont>Back to Table of Contents</a>

---


#### **5.1 Duplicates within the dataset.**
- We only explore the important dataset which is train, test, tags, imdb and movies.

**Comment:** From the above results we can see that they no duplicated values within our dataset.

#### **6.2 Processing the data to remove characters.**
- A function to remove specific character and replace is my an empty space. 

**Comment:** This function will remove character '|' when applied to a dataset with a name movies and have a column called genres. 

<a id="six"></a>
## **6. Modelling**
<a class="anchor" id="1.1"></a>
<a href=#cont>Back to Table of Contents</a>

---


## **Authors**

| Name | Surname | Position |
| :----------- | :------------: | ------------: |
| Percy  | Mmutle       | None       |
|  Lesego  | 88888888      | Project Maneger       |
| Aphiwe | 888888      | None   |
| Tonia | 88888 | None|
|Ntsako| 888888 | None |
| Tumi | 888888 | None |
| Victoria | 888888 | Team Lead |
| Ndivho | 888888 | None |